File: 309.最佳买卖股票时机含冷冻期.c
/*
 * @lc app=leetcode.cn id=309 lang=c
 *
 * [309] 最佳买卖股票时机含冷冻期
 */

// @lc code=start

int maxV(int a, int b)
{
  return a > b ? a : b;
}

int maxProfit(int *prices, int pricesSize)
{
  int i, j, **dp;
  dp = (int **)malloc(sizeof(int *) * pricesSize);
  for (i = 0; i < pricesSize; i++)
    dp[i] = (int *)malloc(sizeof(int) * 4);
  for (i = 0; i < pricesSize; i++)
    for (j = 0; j < 4; j++)
      dp[i][j] = 0;

  dp[0][0] = -prices[0];
  for (i = 1; i < pricesSize; i++)
  {
    dp[i][0] = maxV(maxV(dp[i - 1][0], dp[i - 1][3] - prices[i]), dp[i - 1][1] - prices[i]);
    dp[i][1] = maxV(dp[i - 1][1], dp[i - 1][3]);
    dp[i][2] = dp[i - 1][0] + prices[i];
    dp[i][3] = dp[i - 1][2];
  }

  return maxV(maxV(dp[pricesSize - 1][1], dp[pricesSize - 1][2]), dp[pricesSize - 1][3]);
}
// @lc code=end

File: store/309.最佳买卖股票时机含冷冻期.js
/*
 * @lc app=leetcode.cn id=309 lang=javascript
 *
 * [309] 最佳买卖股票时机含冷冻期
 */

// @lc code=start
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (prices) {
  let dp = new Array(2).fill().map(() => Array(4).fill(0));
  dp[0][0] = -prices[0];
  dp[0][2] = -prices[0];

  for (let i = 1; i < prices.length; i++) {
    dp[1][0] = Math.max(dp[0][0], dp[0][1] - prices[i], dp[0][3] - prices[i]);
    dp[1][1] = Math.max(dp[0][1], dp[0][3]);
    dp[1][2] = Math.max(dp[0][2], dp[0][0] + prices[i]);
    dp[1][3] = Math.max(dp[0][3], dp[0][2]);
    dp[0] = [...dp[1]];
  }

  return dp[1][2] < 0 ? 0 : dp[1][2];
};
// @lc code=end
